File: src/main/java/org/nexial/core/tms/tools/CloseTestRun.java
package org.nexial.core.tms.tools;

import java.util.ArrayList;
import java.util.List;
import org.apache.commons.cli.CommandLine;
import org.apache.commons.cli.DefaultParser;
import org.apache.commons.cli.Options;
import org.apache.commons.cli.ParseException;
import org.apache.commons.collections4.CollectionUtils;
import org.apache.commons.lang3.StringUtils;
import org.jetbrains.annotations.NotNull;
import org.json.simple.JSONArray;
import org.json.simple.JSONObject;
import org.nexial.core.tms.model.TestFile;
import org.nexial.core.tms.model.TmsTestFile;
import org.nexial.core.tms.spi.testrail.TestRailOperations;
import org.nexial.core.utils.InputFileUtils;

import static org.nexial.core.tms.TmsConst.*;
import static org.nexial.core.tms.spi.TmsMetaJson.*;

/**
 * Closes all test runs for the suite associated with the Nexial test file passed in as the argument
 */
public class CloseTestRun {

    public static void main(String[] args) throws ParseException {
        CommandLine cmd = new DefaultParser().parse(initCmdOptions(), args);

        String filepath = "";
        String subplan = "";

        if (cmd.hasOption(SCRIPT) && cmd.hasOption(PLAN)) {
            System.out.println("Only one type of test file is allowed per input");
            System.exit(-1);
        }
        if (cmd.hasOption(SCRIPT)) {
            filepath = cmd.getOptionValue(SCRIPT);
            if (InputFileUtils.isValidScript(filepath)) {
                System.out.println("Invalid test script - " + filepath);
                System.exit(-1);
            }
        } else if (cmd.hasOption(PLAN) && cmd.hasOption(SUBPLAN)) {
            filepath = cmd.getOptionValue(PLAN);
            subplan  = cmd.getOptionValue(SUBPLAN);
            if (!InputFileUtils.isValidPlanFile(filepath)) {
                System.out.println(
                        "specified test plan (" + filepath + ") is not readable or does not contain valid format.");
            }
        } else {
            System.out.println(
                    "The location of the Nexial script/plan file is required. In case of Plan file, subplan is required");
            System.exit(-1);
        }

        if (StringUtils.isNotEmpty(subplan)) {
            System.out.println("Input Plan test file: " + filepath);
            System.out.println("Subplan Name :" + subplan);
        } else {
            System.out.println("Input Script test file: " + filepath);
        }

        // retrieve the project.tms meta file
        TmsTestFile tmsTestFile = getJsonEntryForFile(filepath, subplan);
        TestFile file = tmsTestFile.getFile();
        String suiteId = file.getSuiteId();
        CloseTestRun closeTestRun = new CloseTestRun();
        closeTestRun.closeActiveRuns(suiteId, tmsTestFile.getProjectId());
    }


    /**
     * Command line options
     */
    private static Options initCmdOptions() {
        Options cmdOptions = new Options();
        cmdOptions.addOption(SCRIPT, true, "[REQUIRED] if -" + PLAN + " is missing] The fully qualified " +
                                           "path of the test script.");
        cmdOptions.addOption(PLAN, true, "[REQUIRED if -" + SCRIPT + " is missing] The fully qualified path of a " +
                                         "test plan.");
        cmdOptions.addOption(SUBPLAN, true, "[REQUIRED] if -" + PLAN + "is present.The name of the test plan");
        return cmdOptions;
    }

    /**
     * Close any active test runs associated with the provided suite id
     *
     * @param suiteId the suite id associated with the passed in file
     * @param projectId the project id of the current project
     */
    public void closeActiveRuns(String suiteId, String projectId) {
        TestRailOperations testRail = new TestRailOperations(projectId);
        List<String> activeRunIds = getActiveTestRuns(suiteId, testRail);
        for (String activeRun : activeRunIds) {
            testRail.closeRun(activeRun);
        }
    }

    /**
     * Retrieve the ids of the active test runs associated with the provided suite id
     *
     * @param suiteId the suite id associated with the passed in file
     * @param testRail object of {@link TestRailOperations} class containing the API client
     * @return List of ids of active test runs for the current suite
     */
    @NotNull
    private List<String> getActiveTestRuns(String suiteId, TestRailOperations testRail) {
        List<String> activeRunIds = new ArrayList<>();
        JSONArray activeRuns = testRail.getExistingActiveRuns(suiteId);
        if(CollectionUtils.isEmpty(activeRuns)){
            System.out.println("No active test runs found");
            System.exit(-1);
        }
        System.out.println("Active runs found: ");
        for (Object activeRun : activeRuns) {
            if (activeRun instanceof JSONObject) {
                System.out.println(
                        "Test Run name: " + ((JSONObject) activeRun).get(NAME) + "\tTest Run id: " +
                        ((JSONObject) activeRun).get(ID));
                activeRunIds.add(((JSONObject) activeRun).get(ID).toString());
            }
        }
        return activeRunIds;
    }
}

File: src/main/kotlin/org/nexial/core/tms/tools/CloseTestRun.kt
/*
 * Copyright 2012-2021 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * 	http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package org.nexial.core.tms.tools

import org.apache.commons.cli.DefaultParser
import org.apache.commons.cli.ParseException
import org.apache.commons.collections4.CollectionUtils
import org.nexial.core.tms.spi.TMSOperation
import org.nexial.core.tms.spi.TmsException
import org.nexial.core.tms.spi.TmsFactory
import org.nexial.core.tms.spi.TmsMetaJson
import org.nexial.core.tms.spi.testrail.TestRailOperations
import org.nexial.core.tms.tools.TmsImporter.addCmdOptions
import org.nexial.core.tms.tools.TmsImporter.getArgs
import org.nexial.core.utils.ConsoleUtils

/**
 * Closes all test runs for the suite associated with the Nexial test file passed in as the argument
 */
class CloseTestRun {
    /**
     * Close any active test runs associated with the provided suite id
     *
     * @param suiteId the suite id associated with the passed in file
     * @param projectId the project id of the current project
     */
    fun closeActiveRuns(tms: TMSOperation, activeRunIds: MutableList<String>) {
        activeRunIds.forEach {
            try {
                tms.closeRun(it)
            } catch (e: Exception) {
                ConsoleUtils.error(e.message)
            }
        }
    }

    /**
     * Retrieve the ids of the active test runs associated with the provided suite id.
     *
     * @param suiteId the suite id associated with the passed in file
     * @param tms object of [TestRailOperations] class containing the API client
     * @return List of ids of active test runs for the current suite
     */
    private fun getActiveTestRuns(suiteId: String, tms: TMSOperation): MutableList<String> {
        val activeRuns = try {
             tms.getExistingActiveRuns(suiteId)
        } catch (e: Exception) {
            throw TmsException("Unable to retrieve active test runs due to ${e.message}")
        }

        if (CollectionUtils.size(activeRuns) == 0) {
            throw TmsException("No active test runs found.")
        }
        ConsoleUtils.log("Active test runs found.")
        val activeRunIds = activeRuns.map { it.id.toString() }.toMutableList()
        activeRunIds.forEach {  ConsoleUtils.log("Test Run id: $it") }

        return activeRunIds
    }

    companion object {
        @Throws(ParseException::class)
        @JvmStatic
        fun main(args: Array<String>) {
            val cmdArgs = getArgs(DefaultParser().parse(addCmdOptions(), args))

            // retrieve the project.tms meta file
            val tmsTestFile = TmsMetaJson.getJsonEntryForFile(cmdArgs.first, cmdArgs.second)
            val suiteId = tmsTestFile.file!!.suiteId!!
            val tms = TmsFactory.getTmsInstance(tmsTestFile.projectId)
            val closeTestRun = CloseTestRun()
            val activeRunIds = closeTestRun.getActiveTestRuns(suiteId, tms)
            closeTestRun.closeActiveRuns(tms, activeRunIds)
        }
    }
}